File: MyMonitorHub.JavaAgent/src/main/java/mymonitorhub/agent/core/CommandListenerThread.java
package mymonitorhub.agent.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;

/**
 * Listens for inbound MCM connections on the loopback interface.
 * Each accepted connection is handed off to a CommandProcessingThread.
 * Mirrors C# CommandListenerThread.
 */
class CommandListenerThread implements IThreadShutdown {

    private static final Logger log = LoggerFactory.getLogger(CommandListenerThread.class);

    private final int port;
    private volatile boolean stopped = true;
    private ServerSocket serverSocket;
    private Thread thread;

    CommandListenerThread(int port) {
        this.port = port;
    }

    void start() {
        thread = new Thread(new Runnable() {
            @Override
            public void run() {
                doRun();
            }
        });
        thread.setName("cmd-listener");
        thread.setDaemon(true);
        thread.start();
    }

    @Override
    public void stop() {
        stopped = true;
        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (Exception ignored) {}
        if (thread != null) {
            try { thread.join(10000); } catch (InterruptedException ignored) {}
        }
    }

    private void doRun() {
        stopped = false;
        int errorCount = 0;

        while (!stopped) {
            log.info("Command listener Started on Port: {}", port);
            try {
                // Bind to loopback only — same security rationale as C# version.
                serverSocket = new ServerSocket(port, 100, InetAddress.getLoopbackAddress());
                errorCount = 0;

                while (!stopped) {
                    Socket client = serverSocket.accept();
                    CommandProcessingThread handler = new CommandProcessingThread(client);
                    handler.start();
                }
            } catch (Exception e) {
                if (!stopped) {
                    errorCount++;
                    log.error("Command listener error: {}", e.getMessage());
                    if (errorCount >= 2) {
                        stopped = true;
                        break;
                    }
                    try { Thread.sleep(10000); } catch (InterruptedException ignored) {}
                }
            } finally {
                if (serverSocket != null && !serverSocket.isClosed()) {
                    try { serverSocket.close(); } catch (Exception ignored) {}
                }
            }
        }
        log.info("CommandListenerThread has come to an end");
    }
}

File: MyMonitorHub.Agent/MyMonitorHub.Agent.Core/Core/CommandListenerThread.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using Serilog;

namespace MyMonitorHub.Agent.Core
{
    internal sealed class CommandListenerThread : IThreadShutdown
    {
        private static readonly ILogger Logger = Log.ForContext<CommandListenerThread>();

        private readonly int _port;
        private volatile bool _stopped = true;

        private Socket? _listener;
        private static readonly ManualResetEvent AllDone = new ManualResetEvent(false);


        internal CommandListenerThread(int port)
        {
            _port = port;
        }

        private void Run()
        {
            _stopped = false;
            var count = 0;
            while (!_stopped)
            {
                Logger.Information("Command listener Started on Port: " + _port);
                try
                {
                    // Bind to loopback only. This listener is intended for local IPC
                    // (e.g. backup scripts on the same host firing "OK"/"FAIL" events into
                    // the agent's queue). Binding to IPAddress.Any exposed an unauthenticated
                    // cleartext command port to the entire network, letting any reachable
                    // host inject or suppress monitoring events. See security audit finding A-1.
                    var localEndPoint = new IPEndPoint(IPAddress.Loopback, _port);

                    // Create a TCP/IP socket.
                    _listener = new Socket(AddressFamily.InterNetwork,
                        SocketType.Stream, ProtocolType.Tcp);

                    // Bind the socket to the local endpoint and 
                    // listen for incoming connections.

                    _listener.Bind(localEndPoint);
                    _listener.Listen(100);

                    // Start listening for connections.
                    _stopped = false;
                    while (!_stopped)
                    {
                        // Set the event to nonsignaled state.
                        AllDone.Reset();

                        _listener.BeginAccept(AcceptCallback, _listener);

                        // Wait until a connection is made before continuing.
                        AllDone.WaitOne();

                    }
                }
                catch (Exception e)
                {
                    count++;
                    Logger.Error(e, "Command listener error");
                    if (count == 2)
                    {
                        _stopped = true;
                        break;
                    }
                    Thread.Sleep(10000); // wait 10 seconds
                }

            } // while
            Logger.Information("CommandListenerThread has come to an end");
        }

        void AcceptCallback(IAsyncResult ar)
        {
            // Signal the main thread to continue.
            AllDone.Set();

            // Get the socket that handles the client request.
            var listener = (Socket)ar.AsyncState!;
            var handler = listener.EndAccept(ar);

            var commandProcessing = new CommandProcessingThread(handler);
            commandProcessing.Start();
        }


        private Thread? _thread;
        internal void Start()
        {
            _thread = new Thread(Run);
            _thread.Start();
        }

        public void Stop()
        {
            _stopped = true;
            _listener?.Dispose();
            _thread?.Join(10000); // wait a max of 10 seconds to shut this down
        }
    }

    // class
}